python/lsst/daf/butler/direct_query_driver/_query_builder.py

```diff
@@ -0,0 +1,573 @@
+        )
+        if self.postprocessing:
+            for element in self.postprocessing.iter_missing(self.columns):
+                results.fields.add(
+                    ddl.FieldSpec.for_region(
+                        self.joiner.name_shrinker.shrink(
+                            self.columns.get_qualified_name(element.name, "region")
+                        )
+                    )
+                )
+        return results
+
+
+@dataclasses.dataclass
+class QueryJoiner:
+    """A struct used to represent the FROM and WHERE clauses of an
+    under-construction SQL SELECT query.
+
+    This object's methods frequently "consume" ``self``, by either returning
+    it after modification or returning related copy that may share state with
+    the original.  Users should be careful never to use consumed instances, and
+    are recommended to reuse the same variable name to make that hard to do
+    accidentally.
+    """
+
+    db: Database
+    """Object that abstracts over the database engine."""
+
+    from_clause: sqlalchemy.FromClause | None = None
+    """SQLAlchemy representation of the FROM clause.
+
+    This is initialized to `None` but in almost all cases is immediately
+    replaced.
+    """
+
+    where_terms: list[sqlalchemy.ColumnElement[bool]] = dataclasses.field(default_factory=list)
+    """Sequence of WHERE clause terms to be combined with AND."""
+
+    dimension_keys: NonemptyMapping[str, list[sqlalchemy.ColumnElement]] = dataclasses.field(
+        default_factory=lambda: NonemptyMapping(list)
+    )
+    """Mapping of dimension keys included in the FROM clause.
+
+    Nested lists correspond to different tables that have the same dimension
+    key (which should all have equal values for all result rows).
+    """
+
+    fields: NonemptyMapping[str, dict[str, sqlalchemy.ColumnElement[Any]]] = dataclasses.field(
+        default_factory=lambda: NonemptyMapping(dict)
+    )
+    """Mapping of columns that are neither dimension keys nor timespans.
+
+    Inner and outer keys correspond to the "logical table" and "field" pairs
+    that result from iterating over `~.queries.tree.ColumnSet`, with the former
+    either a dimension element name or dataset type name.
+    """
+
+    timespans: dict[str, TimespanDatabaseRepresentation] = dataclasses.field(default_factory=dict)
+    """Mapping of timespan columns.
+
+    Keys are "logical tables" - dimension element names or dataset type names.
+    """
+
+    special: dict[str, sqlalchemy.ColumnElement[Any]] = dataclasses.field(default_factory=dict)
+    """Special columns that are available from the FROM clause and
+    automatically included in the SELECT clause when this joiner is nested
+    within a `QueryBuilder`.
+
+    These columns are not part of the dimension universe and are not associated
+    with a dataset.  They are never returned to users, even if they may be
+    included in raw SQL results.
+    """
+
+    name_shrinker: NameShrinker | None = None
```

Comment:
I wonder if it would be easier to add a `name_shrinker` property to `Database` instead of worrying about creating them and transferring records between them here.